Rustification: Add NAPI-RS transition-rust-backend with capnp se…

greenscientist opened the pull request:
…rialization

This create a new package, transition-rust-backend, which will contains, as the name imply, rust functions for the backend.
It use the NAPI-RS package to create native binding for Node.

The first thing we implement is direct calls to the Json2Capnp code that write and read the Capnp cache files. The
capnp_serialization.rs file add simple NAPI wrapper for each collection object in the Json2Capnp library. The function
for the single object read/write (for Node and Line) were made async to increase performance. (That way they are added
to the internal libuv queue of NodeJs and run in its multiple threads.)

We went from 90 seconds to 30 seconds when regenerating the cache in the app startup.

For the moment, we integrated the calls in the same function that was doing the fetch to the Json2Capnp deamon with a
simple list of if/else to match the right function for each cache type.

The integration of the direct call to capnp serialization can be refactored. We could probably get rid of the whole
Json2Capnp*.ts files and call the Rust functions directly. We can also rework how we handle the cache directory.

Write path tested by running calculation using TrRouting and checking that the file were the same with sha256sum.

Tested the read path by added a read call in the recreateCache function and checking that the results structures
were the same.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added a Rust-based native Cap’n Proto cache read/write package integrated into the workspace.
  * Cache generation/lookup now uses deterministic local Cap’n Proto cache files for supported cache types, with automatic fallback to the prior network approach for others.
  * Expanded build and packaging setup for broader platform targets (including container builds).
* **Bug Fixes**
  * Improved cache consistency and repeatability for supported cache operations.
* **Tests**
  * Added a native integration test validating the Rust-backed export.
* **Documentation**
  * Updated development setup docs to include Rust formatting and linting tooling.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

tahini commented on package.json:
```diff
@@ -7,7 +7,8 @@
         "packages/transition-backend",
-        "packages/transition-frontend"
+        "packages/transition-frontend",
+        "packages/transition-rust-backend"
```
transition-rust-backend demeure un package nodejs? Aussi, ces packages sont exécutés dans l'ordre, si je comprends bien les dépendences, il faudrait qu'il soit avant chaire-lib-backend ici.

tahini commented on packages/transition-rust-backend/.gitignore:
```diff
@@ -0,0 +1,197 @@
+# Created by https://www.toptal.com/developers/gitignore/api/node
+# Edit at https://www.toptal.com/developers/gitignore?templates=node
```
a-t-on beosin d'un fichier .gitignore spécifique pour ce package?

greenscientist replied:
Plus clean de garder les trucs rust ici à mon avis

tahini commented on packages/transition-rust-backend/index.js:
```diff
@@ -0,0 +1,316 @@
+          )
+          try {
+            if (localFileExisted) {
+              nativeBinding = require('./transition-rust-backend.linux-arm64-gnu.node')
```
Où sont/seront tous ces fichiers? Ce sera compilé? `./` ça veut dire que tout se retrouvera dans le root directory du package? Il faudrait idéalement envoyer tous ces fichiers dans un même répertoire de compilation.

greenscientist replied:
Tout ca est auto généré par napi-rs, donc je sais pas à quel point on peut modifier le tout. a voir.

tahini reviewed (commented):
Je sais que c'est en draft, mais j'ai regardé vite fait c'était quoi cette PR. Finalement, c'est un premier exemple de comment faire faire des choses à Rust par nodejs? Donc éventuellement des calculs de tout genre... Et si on a un point d'entrée pour le faire facilement, ce sera plus simple de décider de passer de nouvelles fonctionnalités à Rust.

greenscientist commented:
@coderabbitai review

greenscientist commented on packages/transition-backend/src/services/json2capnp/Json2CapnpService.ts:
```diff
@@ -16,19 +16,72 @@ const fetch = async (url: string, opts) => {
 class Json2CapnpService {
     // TODO Support the cache path directory or remove from method
-    async writeCache(cacheName: string, jsonData: any, _cachePathDirectory?: string) {
+    async writeCache(cacheName: string, jsonData: any, cachePathDirectory?: string) {
```
Not a change of my PR, won't fix for now

greenscientist commented on packages/transition-backend/src/services/json2capnp/Json2CapnpService.ts:
```diff
@@ -16,19 +16,72 @@ const fetch = async (url: string, opts) => {
+import { capnp_serialization } from 'transition-rust-backend';
+
 class Json2CapnpService {
     // TODO Support the cache path directory or remove from method
```
Not a change of my PR, won't fix for now.(Will probably get deleted in next refactoring)

greenscientist commented on Dockerfile:
```diff
@@ -47,6 +48,9 @@ RUN apt-get -y --no-install-recommends install libboost-chrono1.83.0 liblua5.4-0
 COPY --from=chairemobilite/trrouting:latest /usr/local/bin/trRouting /usr/local/bin/
 COPY --from=greenscientist/osrm-backend:trixie /usr/local/bin/osrm-* /usr/local/bin/
 
+# For Rust binding
+RUN apt-get -y --no-install-recommends install cargo rustfmt rust-clippy
```
We use the container also for development support, so tool are required (but will add comment)

greenscientist commented on packages/transition-backend/src/services/json2capnp/Json2CapnpService.ts:
```diff
@@ -37,16 +90,49 @@ class Json2CapnpService {
 
     async readCache(cacheName: string, params = {}) {
         try {
-            const query = new url.URLSearchParams();
-            for (const param in params) {
-                query.append(param, params[param]);
+            // Copied path from TrRoutingProcessManager
+            const cachePathDirectory = `${directoryManager.projectDirectory}/cache/${config.projectShortname}`;
+            let rustResult = {};
+            if (cacheName === 'agencies') {
+                rustResult = JSON.parse(
+                    capnp_serialization.readAgencyCollection(cachePathDirectory + '/agencies.capnpbin')
+                );
+            } else if (cacheName === 'lines') {
+                rustResult = JSON.parse(capnp_serialization.readLineCollection(cachePathDirectory + '/lines.capnpbin'));
+            } else if (cacheName === 'paths') {
+                rustResult = JSON.parse(capnp_serialization.readPathCollection(cachePathDirectory + '/paths.capnpbin'));
+            } else if (cacheName === 'nodes') {
+                rustResult = JSON.parse(capnp_serialization.readNodeCollection(cachePathDirectory + '/nodes.capnpbin'));
+            } else if (cacheName === 'scenarios') {
+                rustResult = JSON.parse(
+                    capnp_serialization.readScenarioCollection(cachePathDirectory + '/scenarios.capnpbin')
+                );
+            } else if (cacheName === 'services') {
+                rustResult = JSON.parse(
+                    capnp_serialization.readServiceCollection(cachePathDirectory + '/services.capnpbin')
+                );
```
In the test, this was not a big performance issue. (each call took about a second). I'll let the other humans comment if they think its required/useful or not

greenscientist commented on packages/transition-backend/src/services/json2capnp/Json2CapnpService.ts:
```diff
@@ -37,16 +90,49 @@ class Json2CapnpService {
+            } else if (cacheName === 'line') {
+                rustResult = await capnp_serialization.readLineObject(params['uuid'], cachePathDirectory + '/lines');
+            } else if (cacheName === 'node') {
+                rustResult = await capnp_serialization.readNodeObject(params['uuid'], cachePathDirectory + '/nodes');
```
Did you based you comment on similiarity to the rest of the code or it's based on the actual content of the rust binding?

greenscientist commented on packages/transition-rust-backend/src/capnp_serialization.rs:
```diff
@@ -0,0 +1,250 @@
+    let mut file = File::create(file_path)?;
+    service_collection::write_collection(&json, &mut file)
+      .map_err(|e| napi::Error::new(napi::Status::GenericFailure, e.to_string()))
+  }
```
Any code change suggestion on how to do this?

greenscientist commented on packages/transition-rust-backend/.npmignore:
```diff
@@ -0,0 +1,13 @@
+.prettierignore
+rustfmt.toml
+yarn.lock
+*.node
```
Give more more details on this, was the file generated by the napi tool wrong?

greenscientist commented:
@coderabbitai review

tahini commented on packages/transition-backend/src/services/json2capnp/Json2CapnpService.ts:
```diff
@@ -16,19 +16,72 @@ const fetch = async (url: string, opts) => {
+import { directoryManager } from 'chaire-lib-backend/lib/utils/filesystem/directoryManager';
+import config from 'chaire-lib-backend/lib/config/server.config';
+
+import { capnp_serialization } from 'transition-rust-backend';
```
On importe directement from 'transition-rust-backend'? Pas de path supplémentaire ou quoi que ce soit. D,ailleurs, ce package, il est en rust en entier, right? pas de nodejs, c,est un genre de native language.

greenscientist replied:
Ouais, directement, c'est un répertoire dans le package. C'est donc comme n'importe quel package publié. 

Il y a un peu de JS d'auto-généré pour faire la colle,et des tests de validation en TS, mais les fonctionalités et tout le reste est en rust.

tahini commented on packages/transition-backend/src/services/json2capnp/Json2CapnpService.ts:
```diff
@@ -16,19 +16,72 @@ const fetch = async (url: string, opts) => {
-                body: JSON.stringify(jsonData)
-            });
-            return await response.json();
+            if (!cachePathDirectory) {
```
Il faudrait éviter de muter directement la variable envoyée en paramètre. C'est pas clair ce que ça fait.

greenscientist replied:
Comment on fait pour faire en sorte qu'une variable qui est de type "string | undefined" devient juste "string" alors ?

tahini replied:
On l'assigne à une nouvelle variable qui elle est string seulement. `const actualCachePathDirectory = cachePathDirectory === undefined ? 'chemin par défaut' : cachePathDirectory`. Même si tu n'aimes pas ça, c'est comme ça que ça se fait :P

tahini commented on packages/transition-rust-backend/src/capnp_serialization.rs:
```diff
@@ -0,0 +1,312 @@
+  use std::fs::File;
+  use transition_capnp_data::serialization::*;
+
+  // ===========================================================================
```
Est-ce que ce genre de bloc de commentaire est l'équivalent rust d'une jsdoc? Cette doc ira-t-elle automatiquement avec capnp_serialization. Plus généralement, dois-je considérer que la présence de ce commentaire ici remplace le commentaire que je me serais attendue à voir en haut de la fonction ou module dans un autre language?

greenscientist replied:
Non, c'est pas un truc qui autogénère qqch. Juste que ca décrit le bloc de fonctions qui est après.

tahini replied:
Ok, tant qu'à, peut-être écrire de la doc du genre qui va éventuellement générer quelque chose?

greenscientist replied:
Bon, je ne sais pas exactement quels sont tes attentes ici. Ce bout de commentaire décrit de façon générale les fonctions qui suivent. Ca ne serait pas un truc équivalent à du jsdoc dans aucun cas. 
Tous le code de ce fichier n'est qu'une couche de colle simple qui ne sera pas réutilisé ailleur, donc pas tant besoin de faire une longue description de chacune des fonctions. 
Est-ce que tu t'attends à avoir de la documentation sur de quoi ont l'air les fonctions js qui sont généré? Malheureusement je n'ai rien vu dans napi-rs qui permet d'autogénérer ça. 

Si on veut, on peut se mettre à exiger des docstring pour toutes les fonctions/classes/modules, mais ca va s'appliquer aussi au doc TS :)

Ma question est donc surtout est-ce que les commentaires ne sont pas suffisant ou bien pas clair ? Est-ce qu'il manque quelques choses? Est-ce que je dois mettre les fonctions génériques dans un fichier à part pour les documenter différemment du reste? Normalement ces functions ne sont qu'un truc qui est interne qui n'a pas à être publié dans un API externe.

tahini commented on packages/transition-rust-backend/src/capnp_serialization.rs:
```diff
@@ -0,0 +1,312 @@
+    file_path: String,
+    json_str: String,
+  ) -> AsyncTask<WriteCollectionTask> {
+    write_collection_generic(file_path, json_str, line_collection::write_collection)
```
D'où viennent line_collection::write_collection et autres fonctions un peu partout. C'est magique avec `use transition_capnp_data::serialization::*;`?

greenscientist replied:
Bien, magique. Use c'est l'Équivalnt d'un import ou d'un include.    Mais oui, ces fonctions font partie de la Crate transition_capnp_data

tahini commented on packages/transition-rust-backend/src/lib.rs:
```diff
@@ -0,0 +1,15 @@
+
+mod capnp_serialization;
+
+// This is just an temporary example function that we can expose
```
Maybe add a FIXME here to make sure we know it needs to be fixed

tahini commented on packages/transition-rust-backend/Cargo.lock:
```diff
@@ -0,0 +1,564 @@
+]
+
+[[package]]
+name = "transition-rust-backend"
```
It depends on itself?

greenscientist replied:
Probablement que ca veut juste dire qu'il doit se compiler lui aussi. Don't really know.

tahini replied:
> Probablement que ca veut juste dire qu'il doit se compiler lui aussi. Don't really know.

Mais tu es sûr que ça prend ça? Just checking...

greenscientist replied:
Je vais vérifier, mais d'habitude on trust le lock file qui est généré (C'est la même chose que le yarn.lock)

greenscientist replied:
Après vérification, oui, c'est normal, Cargo met le "package" "main" dans la liste du fichier .lock.  Pour comparaison, si tu regarde le Cargo.lock pour json2capnp, tu vas voir qu'il y a aussi un package json2capnp

tahini commented on packages/transition-rust-backend/Cargo.toml:
```diff
@@ -0,0 +1,22 @@
+[package]
+edition = "2021"
```
2026?

greenscientist replied:
Non, c'est pas l'année, mais quel version du language rust on supporte. (C'Est comme la version du standard C++

tahini commented on packages/transition-rust-backend/rustfmt.toml:
```diff
@@ -0,0 +1,2 @@
+tab_spaces = 2
+edition = "2021"
```
2026?

greenscientist replied:
Non, c'est pas l'année, mais quel version du language rust on supporte. (C'Est comme la version du standard C++

tahini reviewed (commented):
Quelques commentaires, je me rends compte que je ne connais pas le rust pentoute. Ce serait bien d'ajouter de la doc sur comment les futurs développements des fonctionnalités rust devraient se faire. La sérialisation de données utilise un package qu'on avait déjà dans `services`, mais est-ce l'approche qu'on désire? Ie que transition-rust-backend soit un thin middleware pour faire le lien entre les appels de transition et une lib développée ailleurs, donc mettons que la génération des paths passe à rust, va-t-on vouloir ajouter une librairie sous `services`? Ou mettre le code directement dans `transition-rust-backend`? Cette doc pourrait aller dans `CONTRIBUTING.MD`

greenscientist commented:
> Quelques commentaires, je me rends compte que je ne connais pas le rust pentoute. Ce serait bien d'ajouter de la doc sur comment les futurs développements des fonctionnalités rust devraient se faire. La sérialisation de données utilise un package qu'on avait déjà dans `services`, mais est-ce l'approche qu'on désire? Ie que transition-rust-backend soit un thin middleware pour faire le lien entre les appels de transition et une lib développée ailleurs, donc mettons que la génération des paths passe à rust, va-t-on vouloir ajouter une librairie sous `services`? Ou mettre le code directement dans `transition-rust-backend`? Cette doc pourrait aller dans `CONTRIBUTING.MD`

Je ne sais pas encore quelle est la meilleure structure pour gérer tout le code Rust, est-ce des lib séparés comme du dit, ou bien on met tout dans le package directement. Je sais pas encore. Il va falloir en faire encore un peu plus avant d'avoir un bon feeling sur la question. Là, on avait déjà le code qui vivait dans services, donc j'ai juste pris ça, mais je sais pas si on veut garder ce pattern à l'avenir.

tahini commented:
> Je ne sais pas encore quelle est la meilleure structure pour gérer tout le code Rust, est-ce des lib séparés comme du dit, ou bien on met tout dans le package directement. Je sais pas encore. Il va falloir en faire encore un peu plus avant d'avoir un bon feeling sur la question. Là, on avait déjà le code qui vivait dans services, donc j'ai juste pris ça, mais je sais pas si on veut garder ce pattern à l'avenir.

Ok, ça a probablement du sens pour l'instant de toute façon, json2capnp est réellement un module et si traclus-rust arrive ensuite, c'est encore un cas de module séparé. Quand on aura des choses plus intriquées (path generation genre? algo génétique? qui sera l'heureux élu de notre prochain rewrite?), on verra comment on met ça.

D'ailleurs, mettons que l'algo génétique gagne le concours, comment suggères-tu que rust s'intègre avec les workers? Est-ce que le code rust roule dans le thread principal de nodejs?

greenscientist committed:
Rustification: Add NAPI-RS transition-rust-backend with capnp serialization

This create a new package, transition-rust-backend, which will contains, as the name imply, rust functions for the backend.
It use the NAPI-RS package to create native binding for Node.

The first thing we implement is direct calls to the Json2Capnp code that write and read the Capnp cache files. The
capnp_serialization.rs file add simple NAPI wrapper for each collection object in the Json2Capnp library. We made
a generic task wrapper, which can be implement for each data type. All call were made async. (That way they are added
to the internal libuv queue of NodeJs and run in its multiple threads.) But the performance gains for these async call
is mostly seen only for the Node and Line single object calls.

We went from 90 seconds to 30 seconds when regenerating the cache in the app startup.

For the moment, we integrated the calls in the same function that was doing the fetch to the Json2Capnp deamon with a
simple list of if/else to match the right function for each cache type.

The integration of the direct call to capnp serialization can be refactored. We could probably get rid of the whole
Json2Capnp*.ts files and call the Rust functions directly. We can also rework how we handle the cache directory.

Write path tested by running calculation using TrRouting and checking that the file were the same with sha256sum.

Tested the read path by added a read call in the recreateCache function and checking that the results structures
were the same.

Simple unit test was added to test rust call with a simple sum function. This can be replaced by an actual
useful function when we have one. We used the vitest framework instead of jest since we are starting a new
package from scratch, as a future proofing measure. (It's also what napi-rs uses)

greenscientist commented:
> Ok, ça a probablement du sens pour l'instant de toute façon, json2capnp est réellement un module et si traclus-rust arrive ensuite, c'est encore un cas de module séparé. Quand on aura des choses plus intriquées (path generation genre? algo génétique? qui sera l'heureux élu de notre prochain rewrite?), on verra comment on met ça.

On va éventuellement que garder la partie lib de json2capnp et on pourra voir où est-ce qu'on garde ça. Des crate séparé ou non. J'imagine que ca va dépendre de comment on veut les faire évoluer et si on voudrait les utiliser de facon indépendante. 
> 
> D'ailleurs, mettons que l'algo génétique gagne le concours, comment suggères-tu que rust s'intègre avec les workers? Est-ce que le code rust roule dans le thread principal de nodejs?

Rust roule dans le thread de node comme n'importe quel bout de code. Donc si on le roule dans un worker, ca va rouler dans ce fil d'exécution. Donc avec des modules napi-rs, tu n'as pas vriament à te soucier de où ça roule, sauf si tu créé toi même de nouveaux threads. Mais si tu fais ça, c'est ta responsabilité de les reconcillier avec le thread qui a appellé la fonction. 

Pour le prochain target, j'imagine qu'il faudra voir quel est le morceau qui est soit trop compliquer en TS et qui aurait un bénéfice d'un langage vraiment typé ou bien juste les morceaux qui sont présentement un bottleneck, genre le traitement des calculs en batchs. Mais à voir.

greenscientist commented:
@coderabbitai resume

tahini reviewed (approved):
Voir s'il y a lieu de mettre des commentaires sur les méthodes rust qui seront publiques. Mais sinon c'est bon

greenscientist commented on packages/transition-backend/src/services/json2capnp/Json2CapnpService.ts:
```diff
@@ -16,19 +16,73 @@ const fetch = async (url: string, opts) => {
+                return await capnp_serialization.writePathCollection(
+                    cachePathDirectory + '/paths.capnpbin',
+                    JSON.stringify(jsonData)
+                );
+            } else if (cacheName === 'nodes') {
+                return await capnp_serialization.writeNodeCollection(
+                    cachePathDirectory + '/nodes.capnpbin',
+                    JSON.stringify(jsonData)
+                );
+            } else if (cacheName === 'scenarios') {
+                return await capnp_serialization.writeScenarioCollection(
+                    cachePathDirectory + '/scenarios.capnpbin',
+                    JSON.stringify(jsonData)
+                );
+            } else if (cacheName === 'services') {
+                return await capnp_serialization.writeServiceCollection(
+                    cachePathDirectory + '/services.capnpbin',
+                    JSON.stringify(jsonData)
+                );
+            } else if (cacheName === 'line') {
+                return await capnp_serialization.writeLineObject(
+                    cachePathDirectory + '/lines',
+                    JSON.stringify(jsonData)
+                );
+            } else if (cacheName === 'node') {
+                return await capnp_serialization.writeNodeObject(
+                    cachePathDirectory + '/nodes',
+                    JSON.stringify(jsonData)
+                );
+            } else {
+                console.warn(`Using legacy capnp write for ${cacheName}`);
+                const request = `${this.getUrlPrefix()}${cacheName}`;
+                const response = await fetch(request, {
+                    method: 'POST',
+                    headers: {
+                        'Content-Type': 'application/json'
+                    },
+                    body: JSON.stringify(jsonData)
+                });
+                return await response.json();
```
This will be refactored later

greenscientist commented:
@coderabbitai review

greenscientist commented:
@coderabbitai review

kaligrafy commented on packages/transition-backend/src/services/json2capnp/Json2CapnpService.ts:
```diff
@@ -37,16 +91,59 @@ class Json2CapnpService {
 
     async readCache(cacheName: string, params = {}) {
         try {
-            const query = new url.URLSearchParams();
-            for (const param in params) {
-                query.append(param, params[param]);
+            // Copied path from TrRoutingProcessManager
+            const cachePathDirectory = `${directoryManager.projectDirectory}/cache/${config.projectShortname}`;
```
Faudrait voir à ça non?

greenscientist replied:
Non, pas vraiment, car il n'y a rien dans le code base actuel qui utilise les fonctions de Read. Mais je vais ajouter un TODO pour ne pas l'oublié dans le refactoring. (La difficulté de le faire avec le code actuel c'est que ce n'est vraiment pas clair comment est gérer le cachePath dans le reste de l'application, donc ca va être plus facile à faire une fois le code simplifié)

kaligrafy commented on packages/transition-backend/src/services/json2capnp/Json2CapnpService.ts:
```diff
@@ -37,16 +90,53 @@ class Json2CapnpService {
+            } else if (cacheName === 'services') {
+                rustResult = JSON.parse(
+                    capnp_serialization.readServiceCollection(cachePathDirectory + '/services.capnpbin')
+                );
+            } else if (cacheName === 'line') {
+                rustResult = JSON.parse(
+                    await capnp_serialization.readLineObject(params['uuid'], cachePathDirectory + '/lines')
+                );
+            } else if (cacheName === 'node') {
+                rustResult = JSON.parse(
+                    await capnp_serialization.readNodeObject(params['uuid'], cachePathDirectory + '/nodes')
```
idem, à vérifier

greenscientist replied:
Je vais ajouter un TODO pour typer les arguments. (Mais la fonction risque de disparaitre dans le refactoring à venir)

kaligrafy reviewed (approved):
Voir les commentaires du lapin, mais sinon, ça me va pour une première implémentation!

greenscientist commented on packages/transition-backend/package.json:
```diff
@@ -70,7 +70,8 @@
         "yargs": "^18.0.0",
-        "zod": "^3.23.8"
+        "zod": "^3.23.8",
+        "transition-rust-backend": "^1.0.0"        
```
@coderabbitai in the future could you surface that in earlier review, this have been there for a while and you did not mention it before.

greenscientist commented on packages/transition-rust-backend/.yarnrc.yml:
```diff
@@ -0,0 +1 @@
+nodeLinker: node-modules
```
@coderabbitai what is the impact of this ?

greenscientist replied:
@coderabbitai please create an issue to validate and follow up on this. Add info on yarn version that would be impacted
